probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -0,0 +1,815 @@
+            ProtocolVersion::V0 => {
+                return Err(ArmError::Probe(
+                    DebugProbeError::CommandNotSupportedByProbe {
+                        command_name: "adiv5 raw ap write",
+                    },
+                ));
+            }
+            ProtocolVersion::V0P => RemoteCommand::WriteApV0P { apsel, addr, value },
+            ProtocolVersion::V1 | ProtocolVersion::V2 => RemoteCommand::WriteApV1 {
+                index,
+                apsel,
+                addr,
+                value,
+            },
+            ProtocolVersion::V3 | ProtocolVersion::V4 => RemoteCommand::WriteApV3 {
+                index,
+                apsel,
+                addr,
+                value,
+            },
+        };
+
+        let result = self
+            .probe
+            .command(command)
+            .map_err(|e| ArmError::Probe(e.into()))?
+            .0;
+        if result == 0 {
+            Ok(())
+        } else {
+            Err(ArmError::Probe(DebugProbeError::Other(format!(
+                "probe returned unexpected result: {}",
+                result
+            ))))
+        }
+    }
+}
+
+impl ArmMemoryInterface for BlackMagicProbeMemoryInterface<'_> {
+    fn ap(&mut self) -> &mut MemoryAp {
+        &mut self.current_ap
+    }
+
+    fn base_address(&mut self) -> Result<u64, ArmError> {
+        self.current_ap.base_address(self.probe)
+    }
+
+    fn get_arm_communication_interface(
+        &mut self,
+    ) -> Result<
+        &mut crate::architecture::arm::ArmCommunicationInterface<
+            crate::architecture::arm::communication_interface::Initialized,
+        >,
+        crate::probe::DebugProbeError,
+    > {
+        Err(DebugProbeError::InterfaceNotAvailable {
+            interface_name: "ARM",
+        })
+    }
+
+    fn try_as_parts(
+        &mut self,
+    ) -> Result<
+        (
+            &mut crate::architecture::arm::ArmCommunicationInterface<
+                crate::architecture::arm::communication_interface::Initialized,
+            >,
+            &mut MemoryAp,
+        ),
+        crate::probe::DebugProbeError,
+    > {
+        Err(DebugProbeError::InterfaceNotAvailable {
+            interface_name: "ARM",
+        })
+    }
+}
+
+impl SwdSequence for BlackMagicProbeMemoryInterface<'_> {
+    fn swj_sequence(
+        &mut self,
+        bit_len: u8,
+        bits: u64,
+    ) -> Result<(), crate::probe::DebugProbeError> {
+        self.probe.swj_sequence(bit_len, bits)
+    }
+
+    fn swj_pins(
+        &mut self,
+        pin_out: u32,
+        pin_select: u32,
+        pin_wait: u32,
+    ) -> Result<u32, crate::probe::DebugProbeError> {
+        self.probe.swj_pins(pin_out, pin_select, pin_wait)
+    }
+}
+
+impl BlackMagicProbeMemoryInterface<'_> {
+    fn read_slice(&mut self, offset: u64, data: &mut [u8]) -> Result<(), ArmError> {
+        // Leave enough room for '&K' plus '#' plus an optional NULL.
```

Comment:
For me as an uneducated person wrt BMP this comment is not fully informative. Could you provide some more info?